libraries/AP_NavEKF3/AP_NavEKF3_core.cpp

```diff
@@ -1417,7 +1422,25 @@ void NavEKF3_core::ConstrainVariances()
     }
 
     if (!inhibitDelVelBiasStates) {
-        for (uint8_t i=13; i<=15; i++) P[i][i] = constrain_float(P[i][i],1e-9f,sq(10.0f * dtEkfAvg));
+        for (uint8_t i=13; i<=15; i++) {
+            if (P[i][i] > 1E-9f) {
+                P[i][i] = fminf(P[i][i], sq(10.0f * dtEkfAvg));
+            } else {
+                // Set the variance to a safe value
+                P[i][i] = 1E-8f;
+                float delVelBiasVar[3];
+                // save the variances
+                for (uint8_t j=13; j<=15; j++) {
+                    delVelBiasVar[j-13] = P[j][j];
+                }
+                // reset covariances
+                zeroCols(P,13,15);
+                // restore variances
+                for (uint8_t j=13; j<=15; j++) {
+                    P[j][j] = delVelBiasVar[j-13];
+                }
```

Comment:
I don't really understand the piece of code above, I think you missed that this is already inside a loop. It looks like either a `break` is missing in the end or this code should simply be:

```
float delVelBiasVar = P[i][i];
zeroCols(P, i, i);
P[i][i] = delVelBiasVar;
```
But in that case I don't know where the `P[i][i] = 1E-8f;` would belong, as that would remove the need to save the variance.